Scripts/Runtime/Data/VirtualData/VirtualData.cs

```diff
@@ -0,0 +1,296 @@
+
+            foreach (IVirtualData source in sources)
+            {
+                virtualData.AddSource(source);
+                source.AddResultDestination(virtualData);
+            }
+
+            return virtualData;
+        }
+
+        private readonly HashSet<IVirtualData> m_Sources = new HashSet<IVirtualData>();
+        private readonly HashSet<IVirtualData> m_ResultDestinations = new HashSet<IVirtualData>();
+        private readonly AccessController m_AccessController;
+
+        private UnsafeTypedStream<TInstance> m_Pending;
+        private DeferredNativeArray<TInstance> m_Iteration;
+        private UnsafeHashMap<TKey, TInstance> m_Lookup;
+
+        /// <summary>
+        /// The underlying <see cref="IDeferredNativeArray"/> used for scheduling
+        /// <see cref="IAnvilJobForDefer"/> or <see cref="IAnvilJobBatchDefer"/> jobs.
+        /// </summary>
+        public DeferredNativeArray<TInstance> ArrayForScheduling
+        {
+            get => m_Iteration;
+        }
+
+        AccessController IVirtualData.AccessController
+        {
+            get => m_AccessController;
+        }
+
+        private VirtualData()
+        {
+            m_AccessController = new AccessController();
+            m_Pending = new UnsafeTypedStream<TInstance>(Allocator.Persistent,
+                                                         Allocator.TempJob);
+            m_Iteration = new DeferredNativeArray<TInstance>(Allocator.Persistent,
+                                                             Allocator.TempJob);
+
+            m_Lookup = new UnsafeHashMap<TKey, TInstance>(MAX_ELEMENTS_PER_CHUNK, Allocator.Persistent);
+        }
+
+        protected override void DisposeSelf()
+        {
+            m_Pending.Dispose();
+            m_Iteration.Dispose();
+            m_Lookup.Dispose();
+
+            RemoveFromSources();
+            m_ResultDestinations.Clear();
+            m_Sources.Clear();
+
+            m_AccessController.Dispose();
+
+            base.DisposeSelf();
+        }
+
+        //*************************************************************************************************************
+        // SERIALIZATION
+        //*************************************************************************************************************
+
+        //TODO: Add Serialization and Deserialization functions to hook into our serialization framework
+
+        //*************************************************************************************************************
+        // RELATIONSHIPS
+        //*************************************************************************************************************
+
+        void IVirtualData.AddResultDestination(IVirtualData resultData)
+        {
+            AddResultDestination(resultData);
+        }
+
+        private void AddResultDestination(IVirtualData resultData)
+        {
+            m_ResultDestinations.Add(resultData);
+        }
```

Comment:
How often do these destinations and sources change? 
I get the impression you'd add/remove a source and destination maybe once per session?

But you'll iterate constantly. Should we instead optimize the collections for iteration? (NativeArrays or List)

I could very well be misunderstanding the use case though

Reply:
Yes you are correct. Switched to `List`